numba/datamodel/models.py

```diff
@@ -0,0 +1,612 @@
+
+    def from_argument(self, builder, value):
+        return self.from_data(builder, value)
+
+    def from_return(self, builder, value):
+        return self.from_data(builder, value)
+
+
+class PrimitiveModel(DataModel):
+    """A primitive type can be represented natively in the target in all
+    usage contexts.
+    """
+
+    def __init__(self, dmm, fe_type, be_type):
+        super(PrimitiveModel, self).__init__(dmm, fe_type)
+        self.be_type = be_type
+
+    def get_value_type(self):
+        return self.be_type
+
+    def as_data(self, builder, value):
+        return value
+
+    def as_argument(self, builder, value):
+        return self.as_data(builder, value)
+
+    def as_return(self, builder, value):
+        return self.as_data(builder, value)
+
+    def from_data(self, builder, value):
+        return value
+
+    def from_argument(self, builder, value):
+        return self.from_data(builder, value)
+
+    def from_return(self, builder, value):
+        return self.from_data(builder, value)
+
+
+@register_default(types.Opaque)
+@register_default(types.NoneType)
+@register_default(types.Function)
+@register_default(types.Type)
+@register_default(types.Object)
+@register_default(types.Module)
+class OpaqueModel(PrimitiveModel):
+    """
+    Passed as opaque pointers
+    """
+
+    def __init__(self, dmm, fe_type):
+        be_type = ir.IntType(8).as_pointer()
+        super(OpaqueModel, self).__init__(dmm, fe_type, be_type)
+
+
+@register_default(types.Integer)
+class IntegerModel(PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        be_type = ir.IntType(fe_type.bitwidth)
+        super(IntegerModel, self).__init__(dmm, fe_type, be_type)
+
+
+class FloatModel(PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        super(FloatModel, self).__init__(dmm, fe_type, ir.FloatType())
+
+
+class DoubleModel(PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        super(DoubleModel, self).__init__(dmm, fe_type, ir.DoubleType())
+
+
+@register_default(types.CPointer)
+class PointerModel(PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        be_type = dmm.lookup(fe_type.dtype).get_data_type().as_pointer()
+        super(PointerModel, self).__init__(dmm, fe_type, be_type)
+
+
+@register_default(types.UniTuple)
+class UniTupleModel(DataModel):
+    def __init__(self, dmm, fe_type):
+        super(UniTupleModel, self).__init__(dmm, fe_type)
+        self._elem_model = dmm.lookup(fe_type.dtype)
+        self._count = len(fe_type)
+
+    def get_value_type(self):
+        elem_type = self._elem_model.get_value_type()
+        return ir.ArrayType(elem_type, self._count)
+
+    def get_data_type(self):
+        elem_type = self._elem_model.get_data_type()
+        return ir.ArrayType(elem_type, self._count)
+
+    def get_return_type(self):
+        return self.get_value_type()
+
+    def get_argument_type(self):
+        return [self._elem_model.get_argument_type()] * self._count
```

Comment:
How about making this a tuple instead of a list? That may simplify some code elsewhere (if sequences are always represented as tuples).